subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/DefaultTypeMetadataStore.java

```diff
@@ -106,19 +106,29 @@ private static String calculateDisplayName(Iterable<? extends PropertyAnnotation
             Annotation normalizationAnnotation = propertyAnnotations.get(NORMALIZATION);
             Class<? extends Annotation> propertyType = determinePropertyType(typeAnnotation, normalizationAnnotation);
             if (propertyType == null) {
-                validationContext.visitPropertyProblem(WARNING,
-                    propertyAnnotationMetadata.getPropertyName(),
-                    String.format("is not annotated with %s", displayName)
+                validationContext.visitPropertyProblem(problem ->
+                    problem.withId(ValidationProblemId.MISSING_ANNOTATION)
+                        .forProperty(propertyAnnotationMetadata.getPropertyName())
+                        .reportAs(ERROR)
+                        .withDescription(() -> "is missing " + displayName)
+                        .happensBecause("A property without annotation isn't considered during up-to-date checking")
+                        .addPossibleSolution(() -> "Add " + displayName)
+                        .addPossibleSolution("Mark it as @Internal")
+                        .documentedAt("validation_problems", "missing_annotation")
                 );
                 continue;
             }
 
             PropertyAnnotationHandler annotationHandler = propertyAnnotationHandlers.get(propertyType);
             if (annotationHandler == null) {
-                validationContext.visitPropertyProblem(ERROR,
-                    propertyAnnotationMetadata.getPropertyName(),
-                    String.format("is annotated with invalid property type @%s",
-                        propertyType.getSimpleName())
+                validationContext.visitPropertyProblem(problem ->
+                    problem.withId(ValidationProblemId.ANNOTATION_INVALID_IN_CONTEXT)
+                        .forProperty(propertyAnnotationMetadata.getPropertyName())
+                        .reportAs(ERROR)
+                        .withDescription(() -> String.format("is annotated with invalid property type @%s", propertyType.getSimpleName()))
+                        .happensBecause(() -> "The '@" + propertyType.getSimpleName() + "' annotation cannot be used in this context")
+                        .addPossibleSolution("Remove the property")
```

Comment:
```suggestion
                        .addPossibleSolution("Remove the property or use a different annotation")
```

I am not sure removing the property is a good suggestion. Did you mean `Remove the annotation`?

Reply:
FYI in case you want to use "or", then you should call `addPossibleSolution`:

```
.addPossibleSolution("Remove the property")
.addPossibleSolution("Use a different annotation")

```